tests/integrationv2/test_early_data.py

```diff
@@ -0,0 +1,495 @@
+@pytest.mark.parametrize("curve", S2N_HRR_CURVES, ids=get_parameter_name)
+@pytest.mark.parametrize("certificate", ALL_TEST_CERTS, ids=get_parameter_name)
+@pytest.mark.parametrize("protocol", [Protocols.TLS13], ids=get_parameter_name)
+@pytest.mark.parametrize("provider", SERVER_PROVIDERS, ids=get_parameter_name)
+@pytest.mark.parametrize("early_data_size", [int(MAX_EARLY_DATA/2), int(MAX_EARLY_DATA-1), MAX_EARLY_DATA, 1])
+def test_s2n_client_with_early_data_rejected_via_hrr(managed_process, tmp_path, cipher, curve, protocol, provider, certificate, early_data_size):
+    if provider == S2N:
+        pytest.skip("S2N does not respect ProviderOptions.curve, so does not trigger a retry")
+
+    early_data_file = str(tmp_path / EARLY_DATA_FILE)
+    early_data = get_early_data_bytes(early_data_file, early_data_size)
+
+    options = ProviderOptions(
+        port=next(available_ports),
+        cipher=cipher,
+        curve=curve,
+        protocol=protocol,
+        insecure=True,
+        use_session_ticket=True,
+        reconnect=True,
+    )
+    options.ticket_file = None
+    options.early_data_file = early_data_file
+    options.max_early_data = MAX_EARLY_DATA
+
+    client_options = copy.copy(options)
+    client_options.mode = Provider.ClientMode
+
+    server_options = copy.copy(options)
+    server_options.mode = Provider.ServerMode
+    server_options.key = certificate.key # Required for the initial connection
+    server_options.cert = certificate.cert # Required for the initial connection
+    server_options.reconnects_before_exit = NUM_CONNECTIONS
+
+    server = managed_process(provider, server_options)
+    s2n_client = managed_process(S2N, client_options)
+
+    for results in s2n_client.get_results():
+        results.assert_success()
+        assert results.stdout.count(to_bytes(S2N_EARLY_DATA_REJECTED_MARKER)) == NUM_RESUMES
+
+    for results in server.get_results():
+        results.assert_success()
+        assert early_data not in results.stdout
+
+
+"""
+Test the S2N server rejecting early data because of a hello retry request.
+
+In order to trigger a successful retry, we need to force the peer to offer us a key share that
+S2N doesn't support while still supporting at least one curve S2N does support.
+"""
+@pytest.mark.uncollect_if(func=invalid_test_parameters)
+@pytest.mark.parametrize("cipher", TLS13_CIPHERS, ids=get_parameter_name)
+@pytest.mark.parametrize("curve", ALL_TEST_CURVES, ids=get_parameter_name)
+@pytest.mark.parametrize("certificate", ALL_TEST_CERTS, ids=get_parameter_name)
+@pytest.mark.parametrize("protocol", [Protocols.TLS13], ids=get_parameter_name)
+@pytest.mark.parametrize("provider", CLIENT_PROVIDERS, ids=get_parameter_name)
+@pytest.mark.parametrize("early_data_size", [int(MAX_EARLY_DATA/2), int(MAX_EARLY_DATA-1), MAX_EARLY_DATA, 1])
+def test_s2n_server_with_early_data_rejected_via_hrr(managed_process, tmp_path, cipher, curve, protocol, provider, certificate, early_data_size):
```

Comment:
what is setting the hello retry request?

Reply:
You don't explicitly request a hello retry in TLS1.3. It's performed whenever the server rejects the offered selection of key shares and asks for a key share for a different supported curve. So you can trigger a HRR by configuring a client to send a key share for a curve the server doesn't support.